File: src/Python/GenerateParentheses.py
def generateParenthesis(n):
    if n == 0:
        return []

    dp = [[] for _ in range(n + 1)]
    dp[0].append("")

    for i in range(1, n + 1):
        for j in range(i):
            for left in dp[j]:
                for right in dp[i - 1 - j]:
                    dp[i].append("(" + left + ")" + right)

    return dp[n]

n = 3
result = generateParenthesis(n)
print(result)

File: src/Java/GenerateParentheses.java
package Java;

import java.util.ArrayList;
import java.util.List;

public class GenerateParentheses {
    public List<String> generateParenthesis(int n) {
        List<List<String>> dp = new ArrayList<>();
        dp.add(new ArrayList<>());
        dp.get(0).add("");

        for (int i = 1; i <= n; i++) {
            List<String> current = new ArrayList<>();
            for (int j = 0; j < i; j++) {
                for (String left : dp.get(j)) {
                    for (String right : dp.get(i - 1 - j)) {
                        current.add("(" + left + ")" + right);
                    }
                }
            }
            dp.add(current);
        }

        return dp.get(n);
    }

    public static void main(String[] args) {
        GenerateParentheses generator = new GenerateParentheses();
        int n = 3;
        List<String> result = generator.generateParenthesis(n);
        System.out.println(result);
    }
}
